File: uva_270_geometry.cpp
/*
"""
Authors: Hung-Hsin Chen <chenhh@par.cse.nsysu.edu.tw>
License: GPL v2

status: AC
difficulty: 1

https://uva.onlinejudge.org/external/2/270.pdf
http://www.algorithmist.com/index.php/UVa_270
*/
#include <iostream>
#include <sstream>
#include <cstdio>
#include <string>
#include <algorithm>
using namespace std;

struct Point {
	int x, y;

	Point() {}

	Point(int _x, int _y){
        x = _x;
        y = _y;
    }

    bool operator < (Point P) const{
        if(x!=P.x)
            return x < P.x;
        return y < P.y;
    }
};

int gcd(int x, int y) {
	if(y == 0)
        return x;
	x = abs(x);
	y = abs(y);
	int tmp;
	while(x%y) {
		tmp = x;
		x = y;
        y = tmp%y;
	}
	return y;
}


int main(){
    int T, n;
    cin>>T;
    string line;
    getline(cin, line);
    getline(cin, line);
    Point data[700];
    Point slope[700];
    while(T--){
        n = 0;
        while(getline(cin, line)){
            if(line=="")
                break;
            sscanf(line.c_str(),"%d %d",&data[n].x,&data[n].y);
            ++n;
        }
        int ans = 2, diff_x, diff_y;
        for(int idx = 0;idx<n-1; ++idx){
            for(int jdx = idx+1, kdx = 0;jdx<n; ++jdx){
                diff_x = data[idx].x-data[jdx].x;
                diff_y = data[idx].y-data[jdx].y;
                if(diff_x || diff_y){
                    int g = __gcd(diff_x, diff_y);
                    diff_x /= g;
                    diff_y /= g;
                }
                slope[kdx] = Point(diff_x, diff_y);
                ++kdx;
            }
            sort(slope,slope+(n-1));
            int aux = 1;
            for(int jdx = 1;jdx+1<n;++jdx){
                if(slope[jdx].x!=slope[jdx-1].x || slope[jdx].y!=slope[jdx-1].y)
                    aux = 1;
                else
                    ++aux;
                ans = max(ans, aux+1);
            }
        }
        cout<<ans<<endl;
        if(T>0)
            cout<<endl;
    }
    return 0;
}

File: uva_270_geometry.py
# -*- coding: utf-8 -*-
"""
Authors: Hung-Hsin Chen <chenhh@par.cse.nsysu.edu.tw>
License: GPL v2

status:
difficulty: 1

https://uva.onlinejudge.org/external/2/270.pdf

3 points in the same line (x1, y1), (x2, y2), (x3, y3)
(y2-y1)/(x2-x1) = (y3-y2)/(x3-x2), then we can get
(y2-y1) * (x3-x2) = (y3-y2)*(x2-x1)

time complexity O(n**3)
"""

import sys


def solve(points):
    # solve
    max_line = 0
    n_points = len(points)
    for idx in range(n_points):
        x1, y1 = points[idx]
        for jdx in range(idx + 1, n_points):
            x2, y2 = points[jdx]
            line = 2
            diff_x, diff_y = (x2 - x1), (y2 - y1)
            for kdx in range(jdx + 1, n_points):
                if kdx != idx and kdx != jdx:
                    x3, y3 = points[kdx]
                    if diff_y * (x3 - x2) == (y3 - y2) * diff_x:
                        line += 1
            if line > max_line:
                max_line = line

    return max_line

def main():
    recs = iter(sys.stdin.readlines())
    points = []
    n_case = int(next(recs))
    _ = next(recs)
    for tdx in range(n_case):
        while 1:
            try:
                point = next(recs).strip()
                if point:
                    points.append(list(map(int, point.split())))
                else:
                    if tdx:
                        print()
                    print(solve(points))
                    points.clear()

            except (StopIteration):
                if len(points):
                    if tdx:
                        print()
                    print(solve(points))
                    points.clear()
                break


if __name__ == '__main__':
    main()
